### PR TITLE
Add helical_geometry

### PR DESCRIPTION
**Pull request by adler-j**
This is work in progress but it seems like a nice feature.

### CODE REVIEW

File: examples/tomo/filtered_backprojection_helical_3d.py [hunk 1]
```diff
@@ -17,29 +17,21 @@
 
 
 # Reconstruction space: discretized functions on the cube
```
Comment (kohr-h) [inline examples/tomo/filtered_backprojection_helical_3d.py]: Actually, we could almost remove this whole line. Better version would be
`# Reconstruction space: functions on [-20, 20]^3, uniformly discretized with 200 samples per dimension`

**Comment by adler-j** [comment]
Now ready for review.

Also note that two rather major bug-fixes are part of this PR, first the parker weighting bug and second the miss-use of pytest.approx that seems to have been quite common.

**Comment by adler-j** [comment]
Bump

File: odl/tomo/geometry/conebeam.py [hunk 2]
```diff
@@ -1435,6 +1435,138 @@ def cone_beam_geometry(space, src_radius, det_radius, num_angles=None,
...
+    >>> geometry.angles.size
+    234
+    >>> geometry.detector.shape
+    (57, 10)
+
+    Notes
+    -----
+    See `cone_beam_geometry` for a description on the mathematics in 2d.
+
+    This geometry adds the Tuy condition.
```
Comment (kohr-h) [inline odl/tomo/geometry/conebeam.py]: Some details?

File: odl/tomo/geometry/conebeam.py [hunk 3]
```diff
@@ -1435,6 +1435,138 @@ def cone_beam_geometry(space, src_radius, det_radius, num_angles=None,
...
+    w = 2 * rho * (rs + rd) / rs
+
+    # Compute minimum number of pixels given the constraint on the
+    # sampling interval and the computed width
+    rb = np.hypot(r, w / 2)  # length of the boundary ray to the flat detector
+    num_px_horiz = 2 * int(np.ceil(w * omega * r / (2 * np.pi * rb))) + 1
+
+    # Compute the vertical size needed in order to get a full sampling
+    # according to the Tuy condition
+    theta = 2 * np.arctan((w / 2) / (src_radius + det_radius))
```
Comment (kohr-h) [inline odl/tomo/geometry/conebeam.py]: Use `rs` and `rd` for consistency

File: odl/tomo/geometry/conebeam.py [hunk 4]
```diff
@@ -1435,6 +1435,138 @@ def cone_beam_geometry(space, src_radius, det_radius, num_angles=None,
...
+
+    # Compute lower and upper bound
+    source_to_line_distance = src_radius * (1 + np.cos(theta))
+    scale = (src_radius + det_radius) / source_to_line_distance
+
+    source_to_line_lower = (pitch *
+                            (num_turns * np.pi / 2 - theta) / (2 * np.pi))
+    h = 2 * source_to_line_lower * scale
```
Comment (kohr-h) [inline odl/tomo/geometry/conebeam.py]: Perhaps 2 more words on the geometric relations used here

**Comment by adler-j** [comment]
Possibility for review on the new version?

File: odl/test/tomo/geometry/geometry_test.py [hunk 5]
```diff
@@ -700,15 +700,51 @@ def test_cone_beam_geometry_helper():
...
+
+    # Create object
+    space = odl.uniform_discr([-1, -1, -2], [1, 1, 2], [20, 20, 40])
+    geometry = odl.tomo.helical_geometry(space, src_radius, det_radius,
+                                         num_turns=num_turns)
+
+    # Validate angles
+    assert geometry.motion_partition.is_uniform
+    assert (geometry.motion_partition.extent ==
+            pytest.approx(num_turns * 2 * np.pi))
```
Comment (olivierverdier) [inline odl/test/tomo/geometry/geometry_test.py]: Same comment: some duplication with `test_cone_beam_geometry_helper` (which is perhaps ok for testing).

File: odl/tomo/geometry/conebeam.py [hunk 6]
```diff
@@ -1435,6 +1435,136 @@ def cone_beam_geometry(space, src_radius, det_radius, num_angles=None,
...
+                         'infinite detector for full coverage')
+    rd = float(det_radius)
+    r = src_radius + det_radius
+    w = 2 * rho * (rs + rd) / rs
+
+    # Compute minimum number of pixels given the constraint on the
+    # sampling interval and the computed width
+    rb = np.hypot(r, w / 2)  # length of the boundary ray to the flat detector
+    num_px_horiz = 2 * int(np.ceil(w * omega * r / (2 * np.pi * rb))) + 1
+
```
[the remark this answers is not in the document]
Comment (adler-j) [reply odl/tomo/geometry/conebeam.py]: I guess it is also shared with `fan_beam_geometry`, I'll see what I can do. Perhaps I'll simply call `fan_beam_geometr` and pick the relevant parameters from it

[the remark this answers is not in the document]
Comment (adler-j) [reply odl/tomo/geometry/conebeam.py]: Ok, I've now read this through and I'm not sure if this is a good idea, it seems I would have to move the 2d code out separately and that part is barely ~10 lines or so. I might actually leave it as is for now

File: odl/test/tomo/geometry/geometry_test.py (hunk 5 above)
Comment (adler-j) [reply odl/test/tomo/geometry/geometry_test.py]: There is quite some duplication but it seems there is also quite some non-duplication :/ separating them in a natural way might be hard so I guess I'll just leave it for now

**Comment by adler-j** [comment]
had a thorough look at the comments and fixed rebasing from master. Can we go ahead with this now?

File: odl/test/tomo/geometry/geometry_test.py (hunk 5 above)
Comment (kohr-h) [reply odl/test/tomo/geometry/geometry_test.py]: Agreed, also with @olivierverdier's comment in parens. Code duplication in tests is more forgivable as long as it's not easy to factor it out into a fixture. The one thing we certainly don't want is a bug in a testing helper 😉

File: odl/tomo/geometry/conebeam.py [hunk 7]
```diff
@@ -1433,6 +1433,158 @@ def cone_beam_geometry(space, src_radius, det_radius, num_angles=None,
...
+    Notes
+    -----
+    In the "fan beam direction", the sampling exactly follows the
+    two-dimensional case see `cone_beam_geometry` for a description.
+
+    In the "axial direction", e.g. along the [0, 0, 1] axis, the geometry is
+    sampled according to two criteria. First, the bounds of the detector
+    are chosen to satisfy the tuy condition.
+    See TSS1998_ for a full description.
```
Comment (kohr-h) [inline odl/tomo/geometry/conebeam.py]: Can you make that a working link? Like
```
`[TSS1998] <link to online resource>`_
```

**Comment by kohr-h** [comment]
Just fix the citation link, other than that I'm good with this version.

**Comment by kohr-h** [comment]
After rebasing and fixing the link this should go in.

**Pull request merged**